### ISSUE TITLE
How do I get jupyterlab to build against my own modded copy of phosphor (for dev/testing purposes)?

### ISSUE THREAD

**Issue by telamonian** [jupyterlab/jupyterlab#5066]
I'm planning to push a couple of PRs upstream to https://github.com/phosphorjs/phosphor with the goal of eventually fixing the contextmenu issues in jupyterlab (#4529, #4590, #4945, etc).

I would, of course, like to test my PRs before I push them. So I figured I would just build jupyterlab against my modded copy of phosphor, only I have no idea how to do that. What's the easiest/simplest workflow for making that happen?

Sorry if this is laid out somewhere in the contrib/dev docs and I just missed missed it.

**Comment by bollwyvl** [comment]
With a checked out repo, you should be able to add your hacked versions to
packages/, and it should get picked up in

jupyter lab build --dev-mode --watch

It might require some additional lerna tinkering

**Comment by jasongrout** [comment]
In your copy of JupyterLab, change the phosphor dependency to point to your local copy. I think you can do this by tinkering with the yarn.lock file (change the "resolved" lines, I believe).

I also just tried this:

1. Delete the packages you want to replace from the `yarn.lock` file - for example the specific phosphor subpackage entry
2. Do `yarn add -W /absolute/path/to/my/phosphor/package` (this adds it to the main package.json and updates the yarn.lock with your package as the pinned dependency)
3. clean the repo (warning this deletes uncommitted files) to get rid of all node_modules directories: `git clean -dfx` - you may want to check with `git status` to see what will be deleted
4. Do `yarn install` to use your own version of the package

I *think* this does what you want...

**Comment by telamonian** [comment]
Okay, I tried out @jasongrout's suggestion, and I was able to do all of the steps, but now when I run `jlpm run build` I get several hundred errors starting with:

```
../../node_modules/@phosphor/application/lib/index.d.ts:3:43 - error TS2307: Cannot find module '@phosphor/widgets'.

3 import { ContextMenu, Menu, Widget } from '@phosphor/widgets';
                                            ~~~~~~~~~~~~~~~~~~~


../../node_modules/@phosphor/datagrid/lib/datagrid.d.ts:2:24 - error TS2307: Cannot find module '@phosphor/widgets'.

2 import { Widget } from '@phosphor/widgets';
                         ~~~~~~~~~~~~~~~~~~~

```

**Comment by telamonian** [comment]
Whoops, I fixed it. I hadn't run `npm run build` in my phosphor dir yet (I did have to install `lerna` to get the build cmd to work).

**Comment by telamonian** [comment]
Alright, so I think I have this totally figured out now. jasongrout's idea works just fine, but it also turns out there's a built-in yarn command, `yarn link`, that does everything that's needed automatically (and faster).

Also, after I linked Jupyterlab to my copy of `@phosphor/widgets` it started behaving in all kinds of strange ways. It turned out that I also needed to link in my copy of `@phosphor/messaging`, even though I hadn't made any changes to that package. It seems that linking in `widgets` alone effectively caused a duplication of the singleton `MessageLoop` that's provided by `messaging`. 

This all seems like something the will come up again/for other people, so I'll add a short section about it to the contrib docs.
